proof/refine/ARM/Schedule_R.thy

```diff
@@ -3368,5 +3317,850 @@ lemma possible_switch_to_corres:
+      apply (fastforce intro: release_queue_modify_tl)
+     apply wpsimp+
+  apply (fastforce simp: st_tcb_at'_def)
+  done
+
+crunches tcb_release_remove
+  for valid_tcbs[wp]: valid_tcbs
+  (wp: crunch_wps update_valid_tcbs simp: crunch_simps)
+
+lemma tcbInReleaseQueue_update_valid_tcb'[wp]:
+  "threadSet (tcbInReleaseQueue_update (\<lambda>_. b)) tcbPtr \<lbrace>valid_tcb' tcb\<rbrace>"
+  apply (wpsimp wp: threadSet_valid_tcb')
+  apply (fastforce simp: valid_tcb'_def tcb_cte_cases_def valid_tcb_state'_def)
+  done
+
+lemma tcbInReleaseQueue_update_valid_tcbs'[wp]:
+  "threadSet (tcbInReleaseQueue_update (\<lambda>_. b)) tcbPtr \<lbrace>valid_tcbs'\<rbrace>"
+  apply (wpsimp wp: threadSet_valid_tcbs')
+  apply (fastforce simp: valid_tcb'_def tcb_cte_cases_def valid_tcb_state'_def)
+  done
+
+lemma tcbReleaseDequeue_valid_tcb'[wp]:
+  "tcbReleaseDequeue \<lbrace>valid_tcb' tcb\<rbrace>"
+  apply (clarsimp simp: tcbReleaseDequeue_def setReleaseQueue_def setReprogramTimer_def)
+  apply ((rule hoare_seq_ext_skip, wpsimp simp: update_valid_tcb') | wpsimp)+
+  done
+
+lemma tcbReleaseDequeue_valid_tcbs'[wp]:
+  "tcbReleaseDequeue \<lbrace>valid_tcbs'\<rbrace>"
+  apply (clarsimp simp: tcbReleaseDequeue_def setReleaseQueue_def setReprogramTimer_def)
+  apply ((rule hoare_seq_ext_skip, wpsimp simp: valid_tcbs'_def update_valid_tcb') | wpsimp)+
+  done
+
+lemma distinct_release_queue_cross_rel:
+  "cross_rel (pspace_aligned and pspace_distinct and valid_release_q)
+             (\<lambda>s'. distinct (ksReleaseQueue s'))"
+  apply (clarsimp simp: cross_rel_def)
+  apply (fastforce dest: state_relation_release_queue_relation
+                 intro!: tcb_at_cross
+                   simp: valid_release_q_def release_queue_relation_def vs_all_heap_simps
+                         obj_at_def is_tcb_def)
+  done
+
+lemma release_queue_nonempty_cross_rel:
+  "cross_rel (pspace_aligned and pspace_distinct and (\<lambda>s. release_queue s \<noteq> []))
+             (\<lambda>s'. ksReleaseQueue s' \<noteq> [])"
+  apply (clarsimp simp: cross_rel_def)
+   apply (fastforce dest: state_relation_release_queue_relation
+                  intro!: tcb_at_cross
+                    simp: valid_release_q_def release_queue_relation_def vs_all_heap_simps
+                          obj_at_def is_tcb_def)
+  done
+
+(* FIXME RT: move? *)
+lemma assert_drop:
+  "\<lbrace>\<lambda>s. P s \<and> Q\<rbrace> assert Q \<lbrace>\<lambda>_. P\<rbrace>"
+  by wpsimp
+
+lemma isRunnable_no_fail[wp]:
+  "no_fail (tcb_at' tcbPtr) (isRunnable tcbPtr)"
+  apply (wpsimp simp: isRunnable_def)
+  done
+
+lemma awaken_body_corres:
+  "corres dc (pspace_aligned and pspace_distinct and valid_objs and valid_sched_action
+              and valid_tcbs and valid_release_q  and (\<lambda>s. Structures_A.release_queue s \<noteq> []))
+             (valid_objs' and valid_queues and valid_queues' and valid_release_queue_iff)
+             Schedule_A.awaken_body awaken_body"
```

Comment:
It took me a while to realise that the function was renamed on the concrete side but the proofs are left stale. A mystery solved... 😅